File: library/src/main/java/me/shkschneider/skeleton/data/CsvWriter.java
package me.shkschneider.skeleton.data;

import java.io.PrintWriter;
import java.io.Writer;

// <http://opencsv.sourceforge.net>
public class CsvWriter {

    private static final char NULL = '\0';

    private PrintWriter mPrintWriter;
    private char mSeparator;
    private char mQuote;
    private char mEscape;
    private String mEol;

    public CsvWriter(final Writer writer) {
        this(writer, ',', '"', '"', "\n");
    }

    public CsvWriter(final Writer writer, final char separator, final char quote, final char escape, final String eol) {
        mPrintWriter = new PrintWriter(writer);
        mSeparator = separator;
        mQuote = quote;
        mEscape = escape;
        mEol = eol;
    }

    public void writeNext(final String nextLine) {
        writeNext(nextLine.split(String.valueOf(mSeparator)));
    }

    public void writeNext(final String[] nextLine) {
        if (nextLine == null) {
            return ;
        }
        final StringBuilder stringBuilder = new StringBuilder();
        for (int i = 0; i < nextLine.length; i++) {
            if (i != 0) {
                stringBuilder.append(mSeparator);
            }
            final String nextElement = nextLine[i];
            if (nextElement == null) {
                continue;
            }
            if (mQuote != NULL) {
                stringBuilder.append(mQuote);
            }
            for (int j = 0; j < nextElement.length(); j++) {
                final char nextChar = nextElement.charAt(j);
                if (mEscape != NULL && nextChar == mQuote) {
                    stringBuilder.append(mEscape).append(nextChar);
                }
                else if (mEscape != NULL && nextChar == mEscape) {
                    stringBuilder.append(mEscape).append(nextChar);
                }
                else {
                    stringBuilder.append(nextChar);
                }
            }
            if (mQuote != NULL) {
                stringBuilder.append(mQuote);
            }
        }
        stringBuilder.append(mEol);
        mPrintWriter.write(stringBuilder.toString());
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/data/CsvWriter.kt
package me.shkschneider.skeleton.data

import java.io.PrintWriter
import java.io.Writer

// <http://opencsv.sourceforge.net>
class CsvWriter {

    private val NULL = '\u0000'
    private val _writer: PrintWriter
    private val _separator: Char
    private val _quote: Char
    private val _escape: Char
    private val _eol: String

    constructor(writer: Writer, separator: Char = ',', quote: Char = '"', escape: Char = '"', eol: String = "\n") {
        _writer = PrintWriter(writer)
        _separator = separator
        _quote = quote
        _escape = escape
        _eol = eol
    }

    fun writeNext(nextLine: String) {
        writeNext(nextLine.split(_separator.toString().toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
    }

    fun writeNext(nextLine: Array<String>?) {
        if (nextLine == null) {
            return
        }
        val stringBuilder = StringBuilder()
        for (i in nextLine.indices) {
            if (i != 0) {
                stringBuilder.append(_separator)
            }
            val nextElement = nextLine[i]
            if (_quote != NULL) {
                stringBuilder.append(_quote)
            }
            (0 until nextElement.length)
                    .map { nextElement[it] }
                    .forEach {
                        if (_escape != NULL && it == _quote) {
                            stringBuilder.append(_escape).append(it)
                        } else if (_escape != NULL && it == _escape) {
                            stringBuilder.append(_escape).append(it)
                        } else {
                            stringBuilder.append(it)
                        }
                    }
            if (_quote != NULL) {
                stringBuilder.append(_quote)
            }
        }
        stringBuilder.append(_eol)
        _writer.write(stringBuilder.toString())
    }

}
